### PR TITLE
network: tc: introduce sfq and tbf

### PR DESCRIPTION
**Pull request by ssahani**
Follow up feature of tc after https://github.com/systemd/systemd/pull/14072 get merged.

### CODE REVIEW

**Comment by yuwata** [comment]
@ssahani I've added tiny commits on top of your commit. PTAL.

**Comment by ssahani** [comment]
Thank you . That was indeed missing. LGTM. Let the CI pass.

**Comment by yuwata** [comment]
Thanks.

How about closing #14072? Then, we can merge this directly.

**Comment by ssahani** [comment]
Yes closed.

**Comment by yuwata** [comment]
Thanks.

**Comment by ssahani** [comment]
CI failure not related.
```
udev                 PASS
root-unittests       PASS
upstream             FAIL timed out
boot-smoke           PASS
```

File: man/systemd.network.xml [hunk 1]
```diff
@@ -2359,6 +2359,32 @@
         </listitem>
       </varlistentry>
 
+      <varlistentry>
+        <term><varname>TokenBufferFilterLatencySec=</varname></term>
+        <listitem>
+          <para>Specifies the latency parameter, which specifies the maximum amount of time a
+          packet can sit in the TBF. The latter calculation takes into account the size of the bucket,
```
Comment (keszybz) [inline man/systemd.network.xml]: Please expand "TBF".

I don't understand the second sentence. The latency is a time. What does the size of the bucket have to  do this? If this is talking about the heuristics of implemention, then that shouldn't be documented.

File: man/systemd.network.xml [hunk 2]
```diff
@@ -2359,6 +2359,32 @@
...
+          the rate and possibly the peakrate (if set). Defaults to unset.</para>
+        </listitem>
+      </varlistentry>
+
+      <varlistentry>
+        <term><varname>TokenBufferFilterBurst=</varname></term>
+        <listitem>
+          <para>Specifies the size of the bucket. This is the maximum amount of bytes that tokens
```
Comment (keszybz) [inline man/systemd.network.xml:2382]: Something's wrong with the second sentence.

File: man/systemd.network.xml [hunk 3]
```diff
@@ -2359,6 +2359,32 @@
...
+          can be available for instantaneously suffixed with K, M, or G, the specified bandwidth is parsed
+          as Kilobytes, Megabytes, or Gigabytes, respectively, to the base of 1000. Defaults to unset.</para>
+        </listitem>
+      </varlistentry>
+
+      <varlistentry>
+        <term><varname>TokenBufferFilterRate=</varname></term>
+        <listitem>
+          <para>Specifies the device specific bandwidth (speedknob) suffixed with K, M, or G, the specified bandwidth is parsed
```
Comment (keszybz) [inline man/systemd.network.xml]: "speedknob" — that is not a word. Please just remove it.

**Comment by yuwata** [comment]
@keszybz Thank you for the comments. The man page is updated. PTAL.

File: man/systemd.network.xml [hunk 4]
```diff
@@ -2368,6 +2368,31 @@
...
+          <para>Specifies the latency parameter, which specifies the maximum amount of time a
+          packet can sit in the Token Buffer Filter (TBF). Defaults to unset.</para>
+        </listitem>
+      </varlistentry>
+
+      <varlistentry>
+        <term><varname>TokenBufferFilterBurst=</varname></term>
+        <listitem>
+          <para>Specifies the size of the bucket. This is the maximum amount of bytes that tokens
+          can be available for instantaneously. When the size is suffixed with K, M, or G, it is parsed
```
Comment (keszybz) [inline man/systemd.network.xml]: for instantaneous transfer?

File: man/systemd.network.xml [hunk 5]
```diff
@@ -2368,6 +2368,31 @@
...
+          as Kilobytes, Megabytes, or Gigabytes, respectively, to the base of 1000. Defaults to unset.</para>
+        </listitem>
+      </varlistentry>
+
+      <varlistentry>
+        <term><varname>TokenBufferFilterRate=</varname></term>
+        <listitem>
+          <para>Specifies the device specific bandwidth suffixed with K, M, or G, the specified bandwidth is parsed
```
Comment (keszybz) [inline man/systemd.network.xml]: Specifies the device specific bandwidth. When suffixed ...

File: src/network/networkd-link.c [hunk 6]
```diff
@@ -2601,7 +2601,7 @@ static int link_configure_qdiscs(Link *link) {
         if (link->qdisc_messages == 0)
                 link->qdiscs_configured = true;
         else
-                log_link_debug(link, "Configuring QDiscs");
+                log_link_debug(link, "Configuring QDisc");
```
Comment (keszybz) [inline src/network/networkd-link.c]: Maybe even "queuing discipline (qdisc)" ?

File: src/network/tc/netem.c [hunk 7]
```diff
@@ -34,33 +34,33 @@ int network_emulator_new(NetworkEmulator **ret) {
         return 0;
 }
 
-int network_emulator_fill_message(Link *link, QDisc *qdisc, sd_netlink_message *req) {
+int network_emulator_fill_message(Link *link, NetworkEmulator *ne, sd_netlink_message *req) {
```
Comment (keszybz) [inline src/network/tc/netem.c]: `const NetworkEmulator *ne` ?

**Comment by yuwata** [comment]
@keszybz Thank you for the review. I hope all comments are addressed now. PTAL.

**Comment by k0ste** [comment]
`htb` is next?

**Pull request merged**